README.md
# @vrbo/fpsmeter

[![NPM Version](https://img.shields.io/npm/v/@vrbo/fpsmeter.svg?style=flat-square)](https://www.npmjs.com/package/@vrbo/fpsmeter)
[![Build Status](https://travis-ci.org/expediagroup/fpsmeter.svg?branch=master)](https://travis-ci.org/expediagroup/fpsmeter)

Optimized javascript utility for measuring frames per second in a browser environment. Useful for observing end-user client run-time performance without adversly impacting performance.

## Installation

```bash
npm install --save @vrbo/fpsmeter
```

## Usage

Within your javascript files, import the component:

```javascript
import FPSMeter from '@vrbo/fpsmeter';
// Configure FPSMeter
const meter = new FPSMeter({
    calculatePerMs: 500, // calculation window for FPS
    onUpdate: (update) => {
        // update.fps - FPS of last window (per defined calculatePerMs option)
        // update.avgfps - FPS average since start()
    },
    onStop: (reason) {
        // reasons why FPSMeter can halt:
        // FPSMeter.stop() initiated by: user
        // FPSMeter.stop() initiated by: document visibilitychange event
        // FPSMeter.stop() initiated by: rAF timed out
        // FPSMeter.stop() initiated by: window blur event
    }
});
// Start
meter.start();
// Stop
meter.stop();
```

Example of collecting 10 FPS calculations of 500ms windows to add to client telemetry summary:

```javascript
import FPSMeter from '@vrbo/fpsmeter';
const meter = new FPSMeter({
    calculatePerMs: 500,
    maxCalculations: 10
});
meter.start();

function onSummarizeClientTelemetry() {
    let summary = {};
    // stop meter
    meter.stop();
    // collect fps windows into string attribute
    summary.fps = meter.fpsWindows.join(',');
    return summary;
}

setTimeout(() => {
    let summary = onSummarizeClientTelemetry();
    console.log(summary.fps); // prints "60,60,60,60,60,60,60,60,60,60" if perfect client performance
}, 10000);
```

\*\* Note: You may see less than 10 FPS measurements if FPSMeter stops for any reason (see below Caveat)

### Caveat: FPSMeter stops

When a user loses focus on the page for any reason, we need to halt the measure of FPS due to the browser engine no longer actively executing the internal API requestAnimationFrame which the FPSMeter library relies on, otherwise our FPS measurements will be skewed.

To mitigate this, FPSMeter will halt when either of the following three conditions occur:

-   FPSMeter.stop() initiated by: document visibilitychange event
-   FPSMeter.stop() initiated by: window blur event
-   FPSMeter.stop() initiated by: rAF timed out 1s (catch all)

This ensures that FPSMeter data is clean and comparable across page views.

To test this for yourself, try the demo link at the top of the page.

## Development

### Starting development harness

```bash
npm start
```

### Prettier

This projects supports auto-formatting of source code! Simply find your favorite IDE from the list in the following list: https://prettier.io/docs/en/editors.html

For VSCode support, perform the following steps:

-   Launch VS Code Quick Open (Ctrl+P)
-   Paste the following command, and press enter:

```
ext install esbenp.prettier-vscode
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0
### Added
-  Initial version

src/index.js
// Copyright 2019 Expedia Group, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export default class FPSMeter {
    constructor({
        onUpdate = () => {},
        onStop = () => {},
        maxFps = 60,
        maxCalculations = null,
        calculatePerMs = 500,
        timeout = 1000
    } = {}) {
        this.calculatePerMs = calculatePerMs;
        this.fpsWindows = [];
        this.maxFps = maxFps;
        this.maxCalculations = maxCalculations;
        this.onStop = onStop;
        this.onUpdate = onUpdate;
        this.timeout = timeout;
        this.stopReason = '';
        this.hasSafeNow = (function(w) {
            const k = 'performance';
            const result = k in w && w[k].now && Date.now() !== w[k].now();
            return result;
        })(window);
    }

    calcfps = (frames, elapsedMs) => {
        return Math.min(parseInt(1000 * (frames / elapsedMs), 10), this.maxFps);
    };

    tick = (timestamp) => {
        if (this.isRunning === false) {
            return false;
        }
        if (!timestamp || !this.hasSafeNow) {
            return this.stop('not-supported');
        }
        if (!this.beginTime) {
            this.beginTime = timestamp;
        }
        if (!this.prevTime) {
            this.prevTime = timestamp;
        }
        const diff = timestamp - this.prevTime;
        this.incFrames++;
        this.totalFrames++;
        if (diff > this.calculatePerMs) {
            // Calculate FPS from previous frame (max 60)
            this.fps = this.calcfps(this.incFrames, diff);
            // Calculate FPS across all frames (max 60)
            this.avgfps = this.calcfps(
                this.totalFrames,
                timestamp - this.beginTime
            );
            // Reset local instances vars
            this.prevTime = timestamp;
            this.incFrames = 0;
            // invoke user callback
            this.publishUpdate();
        }
        if (this.rafTimeout) {
            clearTimeout(this.rafTimeout);
        }
        // Set a timeout for configured `this.timeout` value (defaults to 1s) to expect at least one rAF update
        // Reason why we need to do this is to detect if a user has navigated away from a page
        this.rafTimeout = setTimeout(() => {
            this.stop('timeout');
        }, this.timeout);

        window.requestAnimationFrame(this.tick);
        return true;
    };

    publishUpdate() {
        this.fpsWindows.push(this.fps);
        this.onUpdate({
            fps: this.fps,
            avgfps: this.avgfps
        });
        // stop meter if `this.maxCalculations` is exceeded
        if (this.maxCalculations && this.fpsWindows.length >= this.maxCalculations) {
            this.stop('completed');
        }
    }

    start() {
        if (this.isRunning === true) {
            return false; // no-op
        }
        this.reset();
        // If `window` is undefined, or `requestAnimationFrame` is not available in window scope,
        // mutate to an invalid state, publish invalid update, and return `false`
        if (typeof window !== 'object' || !window.requestAnimationFrame) {
            return false;
        }
        this.isRunning = true;

        this.visibilityListener = document.addEventListener(
            'visibilitychange',
            () => {
                this.stop('visibilitychange');
            }
        );
        this.blurFPSListener = window.addEventListener('blur', () => {
            this.stop('blur');
        });

        window.requestAnimationFrame(this.tick);

        return true;
    }

    stop(reason = 'user') {
        if (this.isRunning === false) {
            return false; // no-op
        }
        window.removeEventListener('visibilitychange', this.visibilityListener);
        window.removeEventListener('blur', this.blurFPSListener);
        if (reason !== 'timeout') {
            clearTimeout(this.rafTimeout);
        }
        this.isRunning = false;
        this.stopReason = reason;
        this.onStop(reason);
        return true;
    }

    reset() {
        this.fpsWindows = [];
        this.fps = 0;
        this.avgfps = 0;
        this.totalFrames = 0;
        this.incFrames = 0;
        this.beginTime = 0;
        this.prevTime = 0;
        this.stopReason = '';
    }
}

/* istanbul ignore next */
if (typeof window !== 'undefined') {
    window.FPSMeter = FPSMeter;
}
